File: felicia/examples/learn/service/ros_srv/python/ros_srv_client_node.py
from functools import partial

import felicia_py as fel
import felicia_py.command_line_interface as cli
from felicia.core.protobuf.master_data_pb2 import ServiceInfo
from felicia.python.rpc.client import Client

from felicia.examples.learn.service.ros_srv._SimpleService import SimpleService, SimpleServiceRequest, SimpleServiceResponse


class RosSrvClientNode(fel.NodeLifecycle):
    def __init__(self, simple_service_flag):
        super().__init__()
        self.simple_service_flag = simple_service_flag
        self.service = simple_service_flag.service_flag.value
        self.client = fel.communication.RosServiceClient(SimpleService)

    def on_did_create(self, node_info):
        self.node_info = node_info
        self.request_register()

    def on_connect(self, service_info_status):
        if (service_info_status == ServiceInfo.REGISTERED):
            self.request_add()
        else:
            print("disconnected...")

    def request_register(self):
        self.client.request_register(
            self.node_info, self.service, self.on_connect)

    def request_unregister(self):
        self.client.request_unregister(
            self.node_info, self.service)

    def on_request_add(self, request, response, status):
        if status.ok():
            print('{} + {} = {}'.format(str(request.a), str(request.b),
                                        cli.TextStyle.green(str(response.sum))))
        else:
            fel.log(fel.ERROR, status.error_message())

    def request_add(self):
        request = SimpleServiceRequest()
        response = SimpleServiceResponse()
        request.a = self.simple_service_flag.a_flag.value
        request.b = self.simple_service_flag.b_flag.value
        self.client.call(request, response, partial(
            self.on_request_add, request, response))

File: felicia/examples/learn/service/ros_srv/cc/ros_srv_client_node.cc
#include "felicia/examples/learn/service/ros_srv/cc/ros_srv_client_node.h"

#include "third_party/chromium/base/strings/string_number_conversions.h"

#include "felicia/core/util/command_line_interface/text_style.h"

namespace felicia {

RosSrvClientNode::RosSrvClientNode(const SimpleServiceFlag& simple_service_flag)
    : simple_service_flag_(simple_service_flag),
      service_(simple_service_flag.service_flag()->value()) {}

void RosSrvClientNode::OnDidCreate(const NodeInfo& node_info) {
  node_info_ = node_info;
  RequestRegister();
}

void RosSrvClientNode::OnConnect(ServiceInfo::Status s) {
  if (s == ServiceInfo::REGISTERED) {
    RequestAdd();
  } else {
    std::cout << "disconnected..." << std::endl;
  }
}

void RosSrvClientNode::RequestRegister() {
  client_.RequestRegister(node_info_, service_,
                          base::BindRepeating(&RosSrvClientNode::OnConnect,
                                              base::Unretained(this)));
}

void RosSrvClientNode::RequestUnregister() {
  client_.RequestUnregister(node_info_, service_);
}

void RosSrvClientNode::OnRequestAdd(const SimpleServiceRequest* request,
                                    SimpleServiceResponse* response,
                                    const Status& s) {
  if (s.ok()) {
    std::cout << base::NumberToString(request->a) << " + "
              << base::NumberToString(request->b) << " = "
              << TextStyle::Green(base::NumberToString(response->sum))
              << std::endl;
  } else {
    LOG(ERROR) << s;
  }
}

void RosSrvClientNode::RequestAdd() {
  SimpleServiceRequest* request = new SimpleServiceRequest();
  SimpleServiceResponse* response = new SimpleServiceResponse();
  request->a = simple_service_flag_.a_flag()->value();
  request->b = simple_service_flag_.b_flag()->value();
  client_->Call(
      request, response,
      base::BindOnce(&RosSrvClientNode::OnRequestAdd, base::Unretained(this),
                     base::Owned(request), base::Owned(response)));
}

}  // namespace felicia